### DISCUSSION TITLE
RenormalizeVector UserObject

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#22185] [Q&A General] [answered]
I have what is probably a question that is easily answered. I am using the UO NormalizeVector that was kindly provided to renormalize a vector field that has to have unit norm. I execute on "timestep_end" (any other option like linear or nonlinear does not work iterations diverge right away)). When I dump data after simulations, there are typically some isolated nodes (using exodus output and paraview) on which the norm deviates somewhat from unity, sometimes up to 2. My question is: does RenormalizeVector renormalize on elements or on nodes? If it normalizes on elements I can understand that the value on nodes can deviate from unity, but if it normalizes on nodes the norm should be unity on all nodes at output (as the normalization is at timestep_end). Or am I misunderstanding something (not at all unlikely....)?

Cheers,
Olle

**Comment by GiudGiud** [comment, marked as answer]
Hello

I dont know this user object. Is it in MOOSE? Could you please send me a link to the source if not. 

It could be a NodalUO or an ElementalUO, I dont know. 
This renormalization on every time step, is it supposed to be conserving something? 

Guillaume

**Comment by heinono1** [comment]
Hi Guillaume,

I think Daniel Schwen wrote it at our request for ferret. The UO is in the
main branch of moose. We use it in micromagnetic simulations as the norm of
the magnetization director has to be conserved.

-Olle

On Sun, Sep 25, 2022 at 3:54 PM Guillaume Giudicelli <
***@***.***> wrote:

> Hello
>
> I dont know this user object. Is it in MOOSE? Could you please send me a
> link to the source if not.
>
> It could be a NodalUO or an ElementalUO, I dont know.
> This renormalization on every time step, is it supposed to be conserving
> something?
>
> Guillaume
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/22185#discussioncomment-3729172>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF76IT6ASUMHP3VBGATWAC3XZANCNFSM6AAAAAAQUKN24Y>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by GiudGiud** [reply]
What’s the exact name of the UO?

**Comment by heinono1** [comment]
RenormalizeVector

On Sun, Sep 25, 2022 at 9:05 PM Guillaume Giudicelli <
***@***.***> wrote:

> What’s the exact name of the UO?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/22185#discussioncomment-3730033>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF3DALAVDIZQN6E5X23WAEAE7ANCNFSM6AAAAAAQUKN24Y>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Comment by heinono1** [comment]
Sorry, the UO is in a sub-branch of Daniel's branch , not the main branch
of moose. I don't know if that one is general accessible for everyone.

On Mon, Sep 26, 2022 at 8:36 AM Olle Heinonen ***@***.***>
wrote:

> RenormalizeVector
>
> On Sun, Sep 25, 2022 at 9:05 PM Guillaume Giudicelli <
> ***@***.***> wrote:
>
>> What’s the exact name of the UO?
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/idaholab/moose/discussions/22185#discussioncomment-3730033>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEKZEF3DALAVDIZQN6E5X23WAEAE7ANCNFSM6AAAAAAQUKN24Y>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>
>
> --
> Olle Heinonen
> ***@***.***
>


-- 
Olle Heinonen
***@***.***

**Comment by mangerij** [comment]
https://github.com/dschwen/moose/blob/renormalize_uo_21801/framework/src/userobject/RenormalizeVector.C

:)

**Reply by GiudGiud** [reply]
so this object loops and normalizes on degrees of freedoms. So quadrature points.

nodes are not quadrature points if you're using linear lagrange. I m surprised it deviates that much, but the normalization is not happening at nodes

**Comment by heinono1** [comment]
Ol, thanks so much, Guillaume.
